WooCommerce: Add variation modal.

justinshreve opened the pull request:
This adds the variation modal that displays when you click the settings cog for a variation in the variation UI (#13698).

It allows you to configure the description and visible status. It also contains a placeholder for the SKU which we can implement once #13653 is handled.

You can also switch between variations within the modal.

Closes #13512.

To test:
* Visit `http://calypso.localhost:3000/store/products/:site/add`
* Start typing in some variation types.
* Click the cog for a variation and see that the modal opens.
* Type in the description box and toggle visible status.
* Switch to another modal using the modal navigation.
* Click done, and click the cog for another variation, to make sure that variation opens instead.

<img width="1280" alt="screen shot 2017-05-09 at 9 13 38 am" src="https://cloud.githubusercontent.com/assets/689165/25861275/1a05032c-3499-11e7-8e25-7373608b4958.png">

@kellychoffman @jameskoster This PR contains UI elements for review.

allendav commented on client/extensions/woocommerce/app/products/product-form-variations-modal.js:
```diff
@@ -0,0 +1,129 @@
+
+		return (
+			<div className="products__product-form-modal-wrapper">
+				<VerticalMenu onClick={ this.switchVariation } className="products__product-form-modal-menu">
```
I'm curious - why don't we have to bind `this` to switchVariation?

justinshreve replied:
We do: https://github.com/Automattic/wp-calypso/pull/13827/files/52cae7654981d26c3bb9383b52bd36e7ca21e480#diff-1d2b47198ed0be90a536ba1cff662b45R34

allendav commented:
At 980px wide, variation dimension input widths are inconsistent

<img width="979" alt="screen shot 2017-05-09 at 12 37 39 pm" src="https://cloud.githubusercontent.com/assets/1595739/25869379/ec8715fe-34b4-11e7-9e85-ccf43730db67.png">

cc @kellychoffman @jameskoster

kellychoffman commented:
> At 980px wide, variation dimension input widths are inconsistent

This is addressed in another PR.

allendav commented on client/extensions/woocommerce/app/products/product-form-variations-modal.js:
```diff
@@ -0,0 +1,129 @@
+							checked={ variation.visible }
+						/>
+					</FormLabel>
+					<p>{ translate(
```
How about using `FormSettingExplanation ` ?

allendav commented on client/extensions/woocommerce/app/products/product-form-variations-modal.js:
```diff
@@ -0,0 +1,129 @@
+					<FormFieldSet className="products__product-form-variation-description">
+						<FormLabel htmlFor="description">{ translate( 'Description' ) }</FormLabel>
+						<FormTextArea name="description" value={ variation.description || '' } onChange={ this.setDescription } />
+						<p>{ translate(
```
How about using `FormSettingExplanation` ?

allendav commented:
The cogs seem to be retaining focus state or something - open the modal by clicking on one, then close the modal and notice the slight difference in shade is retained:

<img width="78" alt="screen shot 2017-05-09 at 1 48 20 pm" src="https://cloud.githubusercontent.com/assets/1595739/25872069/47cd2bf2-34be-11e7-9787-9cc25eef6bc7.png">

cc @kellychoffman @jameskoster

kellychoffman commented:
@allendav: We are removing the cog in favor of linking up the variation name.

justinshreve commented:
@allendav I have updated those strings to use `FormSettingExplanation` (thanks, I didn't know about it before). I have also made updates that remove the cog and link the name per the discussion with Kelly. I've also rebased this, so it should be good for you to continue reviewing.

allendav commented:
Maybe in a later, separate issue, but I think we are going to need debouncing or something. This starts to get a little slow:

<img width="1273" alt="screen shot 2017-05-09 at 3 12 46 pm" src="https://cloud.githubusercontent.com/assets/1595739/25875064/779e5cdc-34ca-11e7-8905-fd7fd8cc0d6f.png">

allendav commented:
If you end up with more than about a dozen variations, the modal vertical list cuts off the rest of them:

<img width="1207" alt="screen shot 2017-05-09 at 3 15 10 pm" src="https://cloud.githubusercontent.com/assets/1595739/25875094/9be854da-34ca-11e7-8fec-55ead6d3a22b.png">

allendav commented on client/extensions/woocommerce/app/products/product-form-variations-modal.js:
```diff
@@ -0,0 +1,132 @@
+	render() {
+		const { variations, translate } = this.props;
+		const { selectedVariation } = this.state;
+		const variation = find( variations, ( v ) => selectedVariation === v.id );
```
Maybe a helper function for finding the currently selected variation? You've used this code 3 places.

allendav commented on client/extensions/woocommerce/app/products/product-form-variations-modal.js:
```diff
@@ -0,0 +1,132 @@
+						/>
+					</FormLabel>
+					<FormSettingExplanation>{ translate(
+						'Hidden variations cannot be selected for purchase by customers.'
```
I wonder if we should only display this warning-like text if they toggle visible to false? cc @kellychoffman - when I see it right now I take an extra moment to think about it and make sure the toggle seems to be enabled

kellychoffman commented:
👍 Styled. The only thing I can't figure out is how to make the modal 600px wide.

jameskoster commented:
> The only thing I can't figure out is how to make the modal 600px wide.

It's because of [this](https://github.com/Automattic/wp-calypso/pull/13604#pullrequestreview-36855287). The modal exists outside of the `.woocommerce` wrapper so all css will just be ignored.

justinshreve commented:
@allendav I have added a helper function for getting the selected variation, and I have added a vertical scroll bar for the menu when there are lots of variations to list. We discussed hiding that help text in Slack, but are going to leave as-is. I am handling debounce in a separate PR since it doesn't relate to these changes. Mind taking another look? Thanks!

allendav commented:
@justinshreve - the modal is coming out way too tall - should probably be limited to the viewport height less some margin? cc @kellychoffman @jameskoster 

<img width="1098" alt="screen shot 2017-05-10 at 8 30 54 am" src="https://cloud.githubusercontent.com/assets/1595739/25907215/b4bcf45c-355b-11e7-9b7b-30c9863f0c77.png">

jameskoster commented:
The variation menu needs a max-height and scrollbar I think.

justinshreve commented:
It does now.

allendav commented:
That works - ty!  There is still some scrolling oddness but I don't think that's fixable. :shipit: 

![and-you-get-a-scrollbar](https://cloud.githubusercontent.com/assets/1595739/25911698/b04e5600-3569-11e7-8f23-9ceaff28b3e4.gif)

justinshreve commented:
Fixed the second modal scrollbar. I don't think there is anything we can do about the background one.
